Dockerfile

```diff
@@ -1,294 +1,550 @@
+ARG DOCKERCLI_VERSION=v17.06.2-ce
+
+# cross compilation helper
+FROM --platform=$BUILDPLATFORM tonistiigi/xx:${XX_VERSION} AS xx
 
-ARG BASE_DEBIAN_DISTRO="bullseye"
-ARG GOLANG_IMAGE="golang:${GO_VERSION}-${BASE_DEBIAN_DISTRO}"
+# go base image to retrieve /usr/local/go
+FROM --platform=$BUILDPLATFORM golang:${GO_VERSION} AS golang
 
-FROM ${GOLANG_IMAGE} AS base
+# dummy stage to make sure the image is built for unsupported deps
+FROM --platform=$BUILDPLATFORM busybox AS build-dummy
+RUN mkdir -p /out
+FROM scratch AS binary-dummy
+COPY --from=build-dummy / /
+
+FROM --platform=$BUILDPLATFORM ${BASE_IMAGE} AS base
+COPY --from=xx / /
 RUN echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache
 ARG APT_MIRROR
 RUN sed -ri "s/(httpredir|deb).debian.org/${APT_MIRROR:-deb.debian.org}/g" /etc/apt/sources.list \
  && sed -ri "s/(security).debian.org/${APT_MIRROR:-security.debian.org}/g" /etc/apt/sources.list
 ENV GO111MODULE=off
-
-FROM base AS criu
 ARG DEBIAN_FRONTEND
-ADD --chmod=0644 https://download.opensuse.org/repositories/devel:/tools:/criu/Debian_11/Release.key /etc/apt/trusted.gpg.d/criu.gpg.asc
-RUN --mount=type=cache,sharing=locked,id=moby-criu-aptlib,target=/var/lib/apt \
-    --mount=type=cache,sharing=locked,id=moby-criu-aptcache,target=/var/cache/apt \
-        echo 'deb https://download.opensuse.org/repositories/devel:/tools:/criu/Debian_11/ /' > /etc/apt/sources.list.d/criu.list \
-        && apt-get update \
-        && apt-get install -y --no-install-recommends criu \
-        && install -D /usr/sbin/criu /build/criu
-
-FROM base AS registry
-WORKDIR /go/src/github.com/docker/distribution
-
-# REGISTRY_VERSION specifies the version of the registry to build and install
-# from the https://github.com/docker/distribution repository. This version of
-# the registry is used to test both schema 1 and schema 2 manifests. Generally,
-# the version specified here should match a current release.
-ARG REGISTRY_VERSION=v2.3.0
-
-# REGISTRY_VERSION_SCHEMA1 specifies the version of the registry to build and
-# install from the https://github.com/docker/distribution repository. This is
-# an older (pre v2.3.0) version of the registry that only supports schema1
-# manifests. This version of the registry is not working on arm64, so installation
-# is skipped on that architecture.
-ARG REGISTRY_VERSION_SCHEMA1=v2.1.0
+RUN --mount=type=cache,sharing=locked,id=moby-base-aptlib,target=/var/lib/apt \
+    --mount=type=cache,sharing=locked,id=moby-base-aptcache,target=/var/cache/apt \
+    apt-get update && apt-get install --no-install-recommends -y \
+      bash \
+      ca-certificates \
+      cmake \
+      file \
+      gcc \
+      git \
+      libc6-dev \
+      lld \
+      make \
+      pkg-config
+COPY --from=golang /usr/local/go /usr/local/go
+ENV GOROOT="/usr/local/go"
+ENV GOPATH="/go"
+ENV PATH="$GOPATH/bin:/usr/local/go/bin:$PATH"
+RUN mkdir -p "$GOPATH/src" "$GOPATH/bin" && chmod -R 777 "$GOPATH"
```

Comment:
This is a follow-up of https://github.com/moby/moby/pull/43553#issuecomment-1116067966. Basically we would need another base image to support the `riscv64` architecture: `ubuntu:22.04`.

As `golang` does not support an ubuntu flavor we need to make sure the ubuntu and debian base images coexist in the Dockerfile otherwise we can't build anymore against `armel`.

This has been tested in https://github.com/crazy-max/moby/blob/a4990cbf40eb4be78d704feb368e9d978756bf6e/Dockerfile#L3-L6 so we can reuse the same stage without changes in https://github.com/crazy-max/moby/blob/a4990cbf40eb4be78d704feb368e9d978756bf6e/Dockerfile#L73-L97.

This is just to avoid drastic changes in a follow-up. Sorry I should have clarify why we would need this.

Reply:
> This is a follow-up of https://github.com/moby/moby/pull/43553#issuecomment-1116067966. Basically we would need another base image to support the riscv64 architecture: ubuntu:22.04.

Ah, hm.. do we know what's missing in Debian? Are there backport / testing debian repositories with packages we need, or is this "newer version of (g)cc etc?"

Trying to see what we could do to simplify things (especially because they're for a still somewhat "esoteric" platform)